serialization.go

```diff
@@ -248,60 +248,104 @@ func IsPreIssuer(issuer *x509.Certificate) bool {
 	return false
 }
 
-// LogEntryFromLeaf converts a LeafEntry object (which has the raw leaf data after JSON parsing)
-// into a LogEntry object (which includes x509.Certificate objects, after TLS and ASN.1 parsing).
-// Note that this function may return a valid LogEntry object and a non-nil error value, when
-// the error indicates a non-fatal parsing error.
-func LogEntryFromLeaf(index int64, leafEntry *LeafEntry) (*LogEntry, error) {
-	var leaf MerkleTreeLeaf
-	if rest, err := tls.Unmarshal(leafEntry.LeafInput, &leaf); err != nil {
-		return nil, fmt.Errorf("failed to unmarshal MerkleTreeLeaf for index %d: %v", index, err)
+// UnmarshaledLeafEntry represents LeafEntry after TLS unmarshaling, or before
+// TLS marshaling.
+type UnmarshaledLeafEntry struct {
+	// Leaf is a MerkleTreeLeaf corresponding to LeafEntry.LeafInput.
+	Leaf MerkleTreeLeaf
+	// Cert is either certificate or pre-certificate.
+	Cert ASN1Cert
+	// Chain is the issuing certificate chain, starting with the issuer of Cert.
+	Chain []ASN1Cert
+	// IsPrecert determines what the Cert field contains. It corresponds to
+	// Leaf.TimestampedEntry.EntryType.
+	IsPrecert bool
```

Comment:
I don't like this as it implies that the entry must be one or the other. It could be an unknown type so would be better to use the correct Go type for it.

Reply:
Could just drop this boolean and use the type indicator, e.g comment `Cert` as:
```
// Cert is:
//  - A certificate if Leaf.TimestampedEntry.EntryType is X509LogEntryType.
//  - A precertificate if Leaf.TimestampedEntry.EntryType is PrecertLogEntryType,
//    in the form of a DER-encoded Certificate as originally added, which includes
//    the poison extension and a signature generated over the pre-cert by the pre-cert
//    issuer).
// - Empty otherwise.
```